x/genutil/types/genesis.go

```diff
@@ -0,0 +1,83 @@
+package types
+
+import (
+	"encoding/json"
+	"os"
+
+	cmtjson "github.com/cometbft/cometbft/libs/json"
+	cmttypes "github.com/cometbft/cometbft/types"
+	"github.com/cosmos/cosmos-sdk/version"
+)
+
+// AppGenesisOnly defines the app's genesis.
+type AppGenesisOnly struct {
+	AppName    string `json:"app_name"`
+	AppVersion string `json:"app_version"`
+}
+
+// AppGenesis defines the app's genesis with the consensus genesis.
+type AppGenesis struct {
+	AppGenesisOnly
+
+	cmttypes.GenesisDoc
+}
+
+func NewAppGenesis(genesisDoc cmttypes.GenesisDoc) *AppGenesis {
+	return &AppGenesis{
+		AppGenesisOnly: AppGenesisOnly{
+			AppName:    version.AppName,
+			AppVersion: version.Version,
+		},
+		GenesisDoc: genesisDoc,
+	}
+}
+
+// SaveAs is a utility method for saving AppGenesis as a JSON file.
+func (ag *AppGenesis) SaveAs(file string) error {
+	// appGenesisBytes, err := ag.MarshalIndent("", "  ")
+	appGenesisBytes, err := json.Marshal(ag)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(file, appGenesisBytes, 0644)
+}
+
+// Marshal the AppGenesis.
+func (ag *AppGenesis) MarshalJSON() ([]byte, error) {
```

Comment:
Is there a way I can unmarshal this in a flatten json?
I have been trying without too much success.

Reply:
Whats happening? App genesis is treating genesis doc as one layer deeper?

The longer term imo is that AppGenesis is able to diverge more significantly from the ABCI genesis doc, so I think at most it really just needs a `.ToAbciGenesisDoc()` function,

Reply:
Yeah, maybe I should go long term directly. The issue right now is I cannot simply `json.Marshal` the struct. CometBFT/TM genesis has its own marshal function (which does not implement the `Marshaller` interface) and does not support the trick of embedding a struct in another to flatten the json. So I need to do that in two steps (which does not work yet).

Reply:
the idea here would then be that we have two genesis files, one for app and one for comet? trying to wrap my head around the issue

Reply:
I think best to to have one genesis. And the app genesis enriches the consensus genesis.

Reply:
One genesis file, but follows the application format.

The application code knows how to convert this app-genesis into a comet-accepted format for the ABCI calls